Pydantic integration does not work with string annotations.

When a project uses `from __future__ import annotations`, all annotations will be stored as strings. This is fairly common, and many projects dictate that any use of type annotations must be accompanied by this import.

The Pydantic integration in Celery introspects the annotations to check if any parameters and/or the return type are subclasses of `pydantic.BaseModel`. This fails when the annotations are `str` instead of the actual class.


I found a mention of this in the thread for another, unrelated issue:

> Same problem here, the dict input is not automatically validated into a Pydantic model:
> 
> ![Image](https://github.com/user-attachments/assets/b1a2cac0-97cb-48d0-90a8-98553b1deffb)
> 
> ```
> AssertionError: Argument must be of type ArgModel, but got {'source_project': 'A
> ```
> Using celery v5.5.0
>  

 _Originally posted by @johannesloibl in [#9442](https://github.com/celery/celery/issues/9442#issuecomment-2782511059)_

--- a/celery/app/base.py
+++ b/celery/app/base.py
@@ -124,14 +124,27 @@ def pydantic_wrapper(
         dump_kwargs = {}
     dump_kwargs.setdefault('mode', 'json')
 
+    # If a file uses `from __future__ import annotations`, all annotations will
+    # be strings. `typing.get_type_hints()` can turn these back into real
+    # types, but can also sometimes fail due to circular imports. Try that
+    # first, and fall back to annotations from `inspect.signature()`.
     task_signature = inspect.signature(task_fun)
+
+    try:
+        type_hints = typing.get_type_hints(task_fun)
+    except (NameError, AttributeError, TypeError):
+        # Fall back to raw annotations from inspect if get_type_hints fails
+        type_hints = None
 
     @functools.wraps(task_fun)
     def wrapper(*task_args, **task_kwargs):
         # Validate task parameters if type hinted as BaseModel
         bound_args = task_signature.bind(*task_args, **task_kwargs)
         for arg_name, arg_value in bound_args.arguments.items():
-            arg_annotation = task_signature.parameters[arg_name].annotation
+            if type_hints and arg_name in type_hints:
+                arg_annotation = type_hints[arg_name]
+            else:
+                arg_annotation = task_signature.parameters[arg_name].annotation
 
             optional_arg = get_optional_arg(arg_annotation)
             if optional_arg is not None and arg_value is not None:
@@ -149,7 +162,11 @@ def pydantic_wrapper(
 
         # Dump Pydantic model if the returned value is an instance of pydantic.BaseModel *and* its
         # class matches the typehint
-        return_annotation = task_signature.return_annotation
+        if type_hints and 'return' in type_hints:
+            return_annotation = type_hints['return']
+        else:
+            return_annotation = task_signature.return_annotation
+
         optional_return_annotation = get_optional_arg(return_annotation)
         if optional_return_annotation is not None:
             return_annotation = optional_return_annotation

--- a/t/integration/tasks.py
+++ b/t/integration/tasks.py
@@ -494,6 +494,13 @@ def add_pydantic(data: AddParameterModel) -> AddResultModel:
     return AddResultModel(result=value)
 
 
+@shared_task(pydantic=True)
+def add_pydantic_string_annotations(data: "AddParameterModel") -> "AddResultModel":
+    """Add two numbers, but with string-annotated Pydantic models (__future__.annotations bug)."""
+    value = data.x + data.y
+    return AddResultModel(result=value)
+
+
 if LEGACY_TASKS_DISABLED:
     class StampOnReplace(StampingVisitor):
         stamp = {"StampOnReplace": "This is the replaced task"}

--- a/t/integration/test_tasks.py
+++ b/t/integration/test_tasks.py
@@ -16,7 +16,7 @@ from celery.worker import state as worker_state
 
 from .conftest import TEST_BACKEND, get_active_redis_channels, get_redis_connection
 from .tasks import (ClassBasedAutoRetryTask, ExpectedException, add, add_ignore_result, add_not_typed, add_pydantic,
-                    fail, fail_unpickleable, print_unicode, retry, retry_once, retry_once_headers,
+                    add_pydantic_string_annotations, fail, fail_unpickleable, print_unicode, retry, retry_once, retry_once_headers,
                     retry_once_priority, retry_unpickleable, return_properties, second_order_replace1, sleeping,
                     soft_time_limit_must_exceed_time_limit)
 
@@ -134,6 +134,20 @@ class test_tasks:
         # Tests calling task only with args
         for i in range(10):
             results.append([i + i, add_pydantic.delay({'x': i, 'y': i})])
+        for expected, result in results:
+            value = result.get(timeout=10)
+            assert value == {'result': expected}
+            assert result.status == 'SUCCESS'
+            assert result.ready() is True
+            assert result.successful() is True
+
+    @flaky
+    def test_pydantic_string_annotations(self, manager):
+        """Tests task call with string-annotated Pydantic model."""
+        results = []
+        # Tests calling task only with args
+        for i in range(10):
+            results.append([i + i, add_pydantic_string_annotations.delay({'x': i, 'y': i})])
         for expected, result in results:
             value = result.get(timeout=10)
             assert value == {'result': expected}

--- a/t/integration/test_tasks.py
+++ b/t/integration/test_tasks.py
@@ -16,9 +16,9 @@ from celery.worker import state as worker_state
 
 from .conftest import TEST_BACKEND, get_active_redis_channels, get_redis_connection
 from .tasks import (ClassBasedAutoRetryTask, ExpectedException, add, add_ignore_result, add_not_typed, add_pydantic,
-                    add_pydantic_string_annotations, fail, fail_unpickleable, print_unicode, retry, retry_once, retry_once_headers,
-                    retry_once_priority, retry_unpickleable, return_properties, second_order_replace1, sleeping,
-                    soft_time_limit_must_exceed_time_limit)
+                    add_pydantic_string_annotations, fail, fail_unpickleable, print_unicode, retry, retry_once,
+                    retry_once_headers, retry_once_priority, retry_unpickleable, return_properties,
+                    second_order_replace1, sleeping, soft_time_limit_must_exceed_time_limit)
 
 TIMEOUT = 10
 

